File: Leetcode/Algorithm/python/Sum of Subarray Minimums.py
INF = 10 ** 10
MOD = (10 ** 9) + 7

class Solution(object):
    def sumSubarrayMins(self, A):
        A.append(-INF)
        st = []
        ans = 0
        for i, num in enumerate(A):
            while st and A[st[-1]] > num:
                cur = st.pop()
                l = cur - (-1 if not st else st[-1])
                r = i - cur
                ans = (ans + l * r * A[cur]) % MOD
            st.append(i)
        return ans

File: Leetcode/Algorithm/cpp/Sum of Subarray Minimums.cc
const int MOD = 1000000000 + 7;

class Solution {
public:
    int sumSubarrayMins(vector<int>& A) {
        set<int> st;
        vector<pair<int, int> > ns;
        const int n = A.size();
        
        for (int i = 0; i < n; i++) {
            ns.push_back({A[i], i});
        }
        sort(ns.begin(), ns.end());
        
        st.insert(-1);
        st.insert(n);
        
        int ans = 0;
        for (const auto& p: ns) {
            int cur = p.second;
            int l = cur - *prev(st.lower_bound(cur));
            int r = *st.upper_bound(cur) - cur;
            // cout << l * r << ' ' << p.first << endl;
            ans = (ans + 1LL * l * r * p.first % MOD) % MOD;
            st.insert(cur);
        }
        return ans;
    }
};
